### PR TITLE
Integrate Async ICQ

### PR DESCRIPTION
**Pull request by nicolaslara**
Closes: #4147 

## What is the purpose of the change

Integrates Async ICQ into osmosis (https://github.com/strangelove-ventures/async-icq)

## Brief Changelog

- Integrates async ICQ into osmosis

## Testing and Verifying

All tests pass

## Documentation and Release Note

  - Does this pull request introduce a new feature or user-facing behavior changes? (yes / no)
  - Is a relevant changelog entry added to the `Unreleased` section in `CHANGELOG.md`? (yes / no)
  - How is the feature or change documented? (not applicable   /   specification (`x/<module>/spec/`)  /  [Osmosis docs repo](https://github.com/osmosis-labs/docs)   /   not documented)

### CODE REVIEW

File: app/keepers/keepers.go [hunk 1]
```diff
@@ -257,6 +262,26 @@ func (appKeepers *AppKeepers) InitNormalKeepers(
 		AddRoute(ibctransfertypes.ModuleName, appKeepers.TransferStack)
 	// Note: the sealing is done after creating wasmd and wiring that up
 
+	// ICQ Keeper
+	icqKeeper := icqkeeper.NewKeeper(
+		appCodec,
+		appKeepers.keys[icqtypes.StoreKey],
+		appKeepers.GetSubspace(icqtypes.ModuleName),
+		appKeepers.IBCKeeper.ChannelKeeper, // may be replaced with middleware
+		appKeepers.IBCKeeper.ChannelKeeper,
```
Comment (ValarDragon) [inline app/keepers/keepers.go:271]: Can we add a comment for what these args are supposed to be? (Confused why theres two)

File: app/keepers/keepers.go [hunk 2]
```diff
@@ -257,6 +262,26 @@ func (appKeepers *AppKeepers) InitNormalKeepers(
...
+		&appKeepers.IBCKeeper.PortKeeper,
+		appKeepers.ScopedICQKeeper,
+		bApp,
```
Comment (ValarDragon) [inline app/keepers/keepers.go:274]: Uhhhh, why does this take in baseApp -- we may have to audit this and all dep updates carefully then. Its not safe to just give base app access to modules.

Comment (nicolaslara) [reply app/keepers/keepers.go:274]: it needs a Querier. I was wondering if there was a something more restrictive that we could pass there.

File: app/keepers/keepers.go (hunk 1 above)
Comment (nicolaslara) [reply app/keepers/keepers.go:271]: Passing it twice is common on IBC modules where you want one with the default implementation and one that is being used to actually route packages through (if there's a middleware, you want to route packets through the middleware but may still need access to the default IBC implementation).

In this case, I don't think the second one is being used. So not sure why it's needed. And the first one is only being used to return the correct app version when implementing `GetAppVersion` (which is required by the interface)

File: app/keepers/keepers.go (hunk 2 above)
Comment (ValarDragon) [reply app/keepers/keepers.go:274]: Gotcha, I think its worth making a wrapper struct on our side for now then

File: app/keepers/keepers.go (hunk 1 above)
Comment (ValarDragon) [reply app/keepers/keepers.go:271]: Oh I see, I imagine this API / standard could be improved, but for another day. SGTM!

File: go.mod [hunk 3]
```diff
@@ -26,6 +26,8 @@ require (
 	github.com/spf13/cobra v1.6.1
 	github.com/spf13/pflag v1.0.5
 	github.com/spf13/viper v1.15.0
+	// Async ICQ branch: ibc-v4
+	github.com/strangelove-ventures/async-icq v0.0.0-20230116084035-5609e84dd443
```
Comment (ValarDragon) [inline go.mod:30]: LGTM, can we add a release blocking TODO for this to be replaced with a tag?

**Review by ValarDragon — APPROVED** [review]
LGTM! Just needs changelog

File: go.mod (hunk 3 above)
Comment (nicolaslara) [reply go.mod:30]: Added https://github.com/osmosis-labs/osmosis/issues/4218. I'll ping them to add a tag

Comment (nicolaslara) [reply go.mod:30]: also, how do I make a ticket release blocking?

Comment (ValarDragon) [reply go.mod:30]: We can add it to the tracker issue for now

**Pull request merged**